### DISCUSSION TITLE
Debugging: "Failed to unmarshal JWT from headers"

### DISCUSSION THREAD

**Discussion by mzgaljic** [kanidm/kanidm#3204] [Q&A] [answered]
Hi, first I want to say, great project! I really appreciate its simplicity and the focus on secure-by-default, and how resource efficient is it. Hard to tick all those boxes in the identity management space. 

Unfortunately I ran into an issue and I'd appreciate some help figuring out what's going on. I think it's a bug but I'm not 100% sure. At the very least, it might be an opportunity to improve the login UX. I searched the discussions and issues and couldn't find any mention of this issue. 🙂

--- Problem Summary ---

After logging into kanidm w/ a passkey on a Mac (safari), i'm greeted with an error page that says `InvalidState`. **I do not see this error page when logging into kanidm on iOS (also safari)**.

--- Steps ---

1. Visit the kanidm dashboard at `https://idm.mydomain.net` (login screen presented), on macOS Safari.
2. I log in with my passkey (Touch ID)
3. I am redirected to `https://idm.mydomain.net/ui/oauth2/resume` which displays:
```
Error
An unrecoverable error occured. Please contact your administrator with the details below.

Error Code: InvalidState

Operation ID: 5af469c7-b0f6-4ce4-89db-5bd39c462bcd
```

---  Error logs (kanidm-server docker container) ---

```
5af469c7-b0f6-4ce4-89db-5bd39c462bcd ERROR    ┕━ 🚨 [error]: unable to resume session, no auth_req was found in the cookie
5af469c7-b0f6-4ce4-89db-5bd39c462bcd ERROR    ┝━ 🚨 [error]: Failed to unmarshal JWT from headers: InvalidSignature
5af469c7-b0f6-4ce4-89db-5bd39c462bcd INFO     request [ 2.01ms | 100.00% ] method: GET | uri: /ui/oauth2/resume | version: HTTP/1.1
```

--- Environment details / architecture ---

Architecture: `Cloudflare w/ DNS proxy → Traefik [port 443] → kanidm container`
Deployment: Traefik and Kanidm are on the same machine, deployed together via docker compose.
Versions: 
- docker images: `traefik:v3.2`, `kanidm/server:1.4.2`
- macOS Sequoia 15.1, M1 Macbook Air

--- Observations / Notes ---

- Refreshing the page on `https://idm.mydomain.net/ui/oauth2/resume` does not fix the issue, I see the same error. 
- In this case as a user i'm redirected to an error page, however, I _am_ successfully logged in! If I manually edit the URL and navigate to `https://idm.mydomain.net`, i'm taken to my list of apps on the kanidm dashboard. I can then successfully access one of my apps (e.g. Grafana via OIDC).
- Logging out and logging back in on my Mac, I see the same error.
- **Manually removing the website data in safari** (settings → privacy → manage website data → click `remove` for `mydomain.net`), **and logging back in, it finally works and I am redirected to my list of apps.**

**Comment by yaleman** [comment]
I'd guess one of your proxy hops is messing with the cookies, maybe take the proxying off cloudflare? Does it work just logging into the UI normally outside an OAuth flow?

**Reply by mzgaljic** [reply]
Thanks for the suggestion! I can't reproduce the error again (after I manually removed the website data in safari settings). I'll keep an eye on it, if it happens again I'll try disabling the Cloudflare DNS proxy. Might also be helpful for me to take a closer look in the browsers network tab.

**Reply by Firstyear** [reply, marked as answer]
This occurs if the servers ephemeral key changes - due to a cookie tampering, server restart, or through a client being directed to a different server by a load balancer. During session setup / auth, we bind things to one server, and so if the client is moved to another server it will break. 

We have just improved some of the errors and diagnostics to help make it clearer to admins what this error is indicating and means to them.
